TestChangeReplicasDescriptorInvariant got stuck

The following test appears to have failed:

[#6620](https://circleci.com/gh/cockroachdb/cockroach/6620):

```
W0828 15:03:57.947676 949 multiraft/multiraft.go:930  node 200000002 failed to send message to 100000001: multiraft/transport.go:157: transport is closed
W0828 15:03:58.028524 949 multiraft/multiraft.go:930  node 200000002 failed to send message to 100000001: multiraft/transport.go:157: transport is closed
W0828 15:03:58.028688 949 multiraft/multiraft.go:930  node 200000002 failed to send message to 300000003: multiraft/transport.go:157: transport is closed
W0828 15:03:58.048768 949 multiraft/multiraft.go:930  node 300000003 failed to send message to 100000001: multiraft/transport.go:157: transport is closed
W0828 15:03:58.048940 949 multiraft/multiraft.go:930  node 300000003 failed to send message to 200000002: multiraft/transport.go:157: transport is closed
panic: test timed out after 5m0s

goroutine 3512 [running]:
testing.startAlarm.func1()
	/tmp/workdir/go/src/testing/testing.go:703 +0x16b
created by time.goFunc
	/tmp/workdir/go/src/time/sleep.go:129 +0x6e

goroutine 1 [chan receive, 4 minutes]:
testing.RunTests(0x14c3bd0, 0x1a38c00, 0xb9, 0xb9, 0x1252001)
	/tmp/workdir/go/src/testing/testing.go:562 +0xafa
--
goroutine 3429 [select]:
github.com/coreos/etcd/raft.(*multiNode).run(0xc8204728a0)
	/go/src/github.com/coreos/etcd/raft/multinode.go:180 +0x2b32
created by github.com/coreos/etcd/raft.StartMultiNode
	/go/src/github.com/coreos/etcd/raft/multinode.go:56 +0x338
FAIL	github.com/cockroachdb/cockroach/storage	300.076s
=== RUN   TestBatchBasics
I0828 14:59:01.153953 959 storage/engine/rocksdb.go:122  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.00s)
=== RUN   TestBatchGet
I0828 14:59:01.156977 959 storage/engine/rocksdb.go:122  closing in-memory rocksdb instance
--- PASS: TestBatchGet (0.00s)
=== RUN   TestBatchMerge
I0828 14:59:01.161901 959 storage/engine/rocksdb.go:122  closing in-memory rocksdb instance
--- PASS: TestBatchMerge (0.00s)
=== RUN   TestBatchProto

```
Please assign, take a look and update the issue accordingly.

--- a/rpc/client.go
+++ b/rpc/client.go
@@ -212,28 +212,29 @@ func (c *Client) runHeartbeat(retryOpts retry.Options, closer <-chan struct{}) {
 		}
 	}
 
-	connErr := errUnstarted // initial condition
-	var beatErr error
+	var err = errUnstarted // initial condition
 	for {
 		for r := retry.Start(retryOpts); r.Next(); {
-			// Reconnect if connection failed or heartbeat error is not
-			// definitely temporary.
-			if netErr, ok := beatErr.(net.Error); connErr != nil || beatErr != nil && !(ok && netErr.Temporary()) {
-				if connErr = c.connect(); connErr != nil {
-					log.Warning(connErr)
+			// Reconnect on failure.
+			if err != nil {
+				if err = c.connect(); err != nil {
 					setUnhealthy()
+					log.Warning(err)
 					continue
 				}
 			}
 
-			if beatErr = c.heartbeat(); beatErr == nil {
-				setHealthy()
-				break
-			} else {
-				log.Warning(beatErr)
+			// Heartbeat regardless of failure.
+			if err = c.heartbeat(); err != nil {
 				setUnhealthy()
+				log.Warning(err)
+				continue
 			}
+
+			setHealthy()
+			break
 		}
+
 		// Wait after the heartbeat so that the first iteration gets a wait-free
 		// heartbeat attempt.
 		select {

--- a/rpc/client_test.go
+++ b/rpc/client_test.go
@@ -25,13 +25,19 @@ import (
 	"github.com/cockroachdb/cockroach/util"
 	"github.com/cockroachdb/cockroach/util/hlc"
 	"github.com/cockroachdb/cockroach/util/leaktest"
+	"github.com/cockroachdb/cockroach/util/retry"
 	"github.com/cockroachdb/cockroach/util/stop"
 )
 
 func init() {
-	// Setting the heartbeat interval in individual tests
-	// triggers the race detector, so this is better for now.
+	// Setting these in individual tests triggers the race detector, so
+	// this is better for now.
+	// TODO(tamird): stop doing this.
 	heartbeatInterval = 10 * time.Millisecond
+	clientRetryOptions = retry.Options{
+		InitialBackoff: 1 * time.Millisecond,
+		MaxBackoff:     1 * time.Millisecond,
+	}
 }
 
 func TestClientHeartbeat(t *testing.T) {

